proposals/002-text-utf-default.md

```diff
@@ -0,0 +1,307 @@
+in UTF-8 encoding. So `text` needs to convert (UTF-8 to UTF-16) all inputs and usually
+ends up converting outputs as well (this time UTF-16 to UTF-8).
+
+Even within Haskell ecosystem UTF-16 is rarely used
+for interprocess communication or as a component of binary formats.
+The very `instance Binary Text` serializes `Text` in UTF-8 encoding.
+
+If we switch the internal representation of `Text` from UTF-16 to UTF-8,
+all such conversions would be made redundant and we'll be able just check that
+a `ByteString` is a valid UTF-8 (which is most often the case) and copy it into `Text`.
+If in future (see an upcoming "Unifying vector-like types" proposal) `ByteString`
+is backed by unpinned memory, we'd be able to eliminate copying entirely.
+
+`Text` is also often used in contexts, which involve mostly ASCII characters.
+This often prompts developers to use `ByteString` instead of `Text` to save 2x space
+in a (false) hope that their data would never happen to contain anything non-ASCII.
+Backing `Text` by UTF-8 removes this source of inefficiency, reduces memory consumption
+up to 2x and promote more convergence to use `Text` for all stringy things
+(as opposed to binary data).
+
+Modern computer science research is focused on developing faster algorithms
+for UTF-8 data, e. g., an ultra-fast JSON decoder [simdjson](https://github.com/simdjson/simdjson). There is much less work on (and demand for) UTF-16 algorithms.
+Switching `text` to UTF-8 will open us a way to accomodate and benefit
+from future developments in rapid text processing.
+
+The importance of UTF-16 to UTF-8 transition was recognised long ago, and at least
+two attempts have been made:
+[in 2011](https://github.com/jaspervdj/text/tree/utf8) and five years later
+[in 2016](https://github.com/text-utf8/text-utf8). Unfortunately, they did not get
+merged into main `text` package. Today, five more years later it seems suitable
+to make another attempt.
+
+# Goals
+
+-   Solicit feedback from community members and industry to gauge appetite for such a change.
+
+-   Provide an implementation, migration, and delivery plan for changing the default encoding of `text` from UTF-16 to UTF-8.
+
+-   Ensure stakeholders (e.g. GHC, Cabal, Stack, boot libs) have ample time to migrate and address any bugs.
+
+-   Performance satisfies targets listed below in "Performance impact" section.
+
+# People
+
+-   Performers:
+
+    -   Leader: Andrew Lelechenko (Bodigrim)
+
+    -   Support: Emily Pillmore (emilypi)
+
+-   Reviewers:
+
+    -   The text maintainers
+
+        -   Xia Li-Yao (lysxia)
+
+        -   Emily Pillmore (emilypi)
+
+        -   Dan Cartwright (chessai)
+
+        -   Callan McGill (boarders)
+
+-   Stakeholders:
+
+    -   Edward Kmett: has been vocal about his use of `text-icu` and requires it not be broken.
+
+-   Ben Gamari: integration with GHC
+
+-   The Cabal maintainers (fgaz, emilypi, mikolaj): integration with Cabal
```

Comment:
Neither `Cabal` nor `cabal-install` use `Text` at the moment. It does use (vendored version) of `ShortText` though.

IMO you can just drop `Cabal` from the list, it won't be affected unless there is some need to use `Text` in `Cabal` (which there isn't as far as I can tell).

Reply:
I don't know much about Cabal internals, but `Cabal.cabal` seems to include `text`:
https://github.com/haskell/cabal/blob/4f8aeb2c8a0a3638e1af887dc869a17e291c8329/Cabal/Cabal.cabal#L272

Reply:
It defines few instances for `Text`, it's not really used.

```
cabal master % git grep 'Data.Text'

-- tests:
Cabal-tests/tests/UnitTests/Distribution/Utils/Generic.hs:import qualified Data.Text as T
Cabal-tests/tests/UnitTests/Distribution/Utils/Generic.hs:import qualified Data.Text.Encoding as T

example (note in comment)
Cabal/src/Distribution/Backpack.hs:-- >>> eitherParsec "foo[Str=text-1.2.3:Data.Text.Text]" :: Either String OpenUnitId
Cabal/src/Distribution/Backpack.hs:-- Right (IndefFullUnitId (ComponentId "foo") (fromList [(ModuleName "Str",OpenModule (DefiniteUnitId (DefUnitId {unDefUnitId = UnitId "text-1.2.3"})) (ModuleName "Data.Text.Text"))]))

-- CharParsing class has
-- text :: Text -> m Text
-- method as it's in upstream parsers -package version
Cabal/src/Distribution/Compat/CharParsing.hs:import Data.Text (Text, unpack)

-- these are used for debug, not part of proper build
Cabal/src/Distribution/Fields/Lexer.hs:import qualified Data.Text   as T
Cabal/src/Distribution/Fields/Lexer.hs:import qualified Data.Text.Encoding as T
Cabal/src/Distribution/Fields/Lexer.hs:import qualified Data.Text.Encoding.Error as T

-- ditto
Cabal/src/Distribution/Fields/LexerMonad.hs:import qualified Data.Text          as T
Cabal/src/Distribution/Fields/LexerMonad.hs:import qualified Data.Text.Encoding as T
Cabal/src/Distribution/Fields/Parser.hs:import qualified Data.Text                as T
Cabal/src/Distribution/Fields/Parser.hs:import qualified Data.Text.Encoding       as T
Cabal/src/Distribution/Fields/Parser.hs:import qualified Data.Text.Encoding.Error as T

-- instances
Cabal/src/Distribution/Utils/Structured.hs:import qualified Data.Text                    as T
Cabal/src/Distribution/Utils/Structured.hs:import qualified Data.Text.Lazy               as LT

-- some dev stuff
bootstrap/src/Main.hs:import qualified Data.Text as T
cabal-dev-scripts/src/GenSPDX.hs:import Data.Text        (Text)
cabal-dev-scripts/src/GenSPDX.hs:import qualified Data.Text            as T
cabal-dev-scripts/src/GenSPDXExc.hs:import Data.Text        (Text)
cabal-dev-scripts/src/GenSPDXExc.hs:import qualified Data.Text            as T
cabal-dev-scripts/src/GenUtils.hs:import Data.Text    (Text)
cabal-dev-scripts/src/GenUtils.hs:import qualified Data.Text           as T

-- this is really fun, the usage is
-- let command' = command { commandName = T.unpack . T.replace "v1-" "" . T.pack . commandName $ command }
cabal-install/src/Distribution/Client/CmdLegacy.hs:import qualified Data.Text as T
cabal-testsuite/src/Test/Cabal/Plan.hs:import qualified Data.Text as Text

-- docs
doc/cabal-package.rst:a requirement ``Str`` and an implementation ``Data.Text``, you can
doc/cabal-package.rst:  ``mixins: parametrized requires (Str as Data.Text)``
doc/cabal-package.rst:  ``mixins: text (Data.Text as Str)``
doc/cabal-package.rst:        parametrized (MyModule as MyModule.Text) requires (Str as Data.Text),

-- lexer debugging (same as above)
templates/Lexer.x:import qualified Data.Text   as T
templates/Lexer.x:import qualified Data.Text.Encoding as T
templates/Lexer.x:import qualified Data.Text.Encoding.Error as T
```

If any change affects `Cabal` performance, that will be *interesting*.

Reply:
Thinking of `Cabal` made me remember `pandoc`. It can be called *the* Haskell app, and it uses `Text` extensively. So if you want to have some real project on the list, i'd nominate it.

The promise that API doesn't change too much should make building `pandoc` even with its dependency footprint only a matter of CPU time.

Reply:
Ah, interesting, thanks.

Reply:
@emilypi shall we remove Cabal from stakeholders?